### PR TITLE
Typescript

### LINKED ISSUE #5: Convert to TS
Convert the project to typescript.

### PR DESCRIPTION
**Pull request by borangoher**
close #5.

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,14 +1,15 @@
...
+import Home from "./components/Home";
+import PaymentForm from "./components/payment-form/PaymentForm";
+import Login from "./components/login/Login";
+import Info from "./components/Info";
+import { LoginProvider } from "./components/LoginContext";
 import "@fontsource/roboto/300.css";
 import "@fontsource/roboto/400.css";
 import "@fontsource/roboto/500.css";
 import "@fontsource/roboto/700.css";
+import React from "react";
```
Comment (yeged) [inline src/App.tsx:12]: We import React on top of file.

File: src/Components/payment-form/PaymentForm.tsx [hunk 2]
```diff
@@ -37,7 +37,7 @@ const PaymentForm = () => {
     resolver: yupResolver(schema),
     defaultValues: paymentFormDefaultValues,
   });
```
Comment (yeged) [inline src/Components/payment-form/PaymentForm.tsx:39]: Can you define form prop types in constant file and import in here to use at useForm. Same for Login.

`useForm<FormProps>`

```
interface FormProps {
    cardholderName: string,
    ...
}
```

File: src/Components/payment-form/PaymentForm.tsx [hunk 3]
```diff
@@ -37,7 +37,7 @@ const PaymentForm = () => {
...
-  const onSubmit = (data) => console.log(data);
+  const onSubmit = (data: object) => console.log(data);
```
Comment (yeged) [inline src/Components/payment-form/PaymentForm.tsx:40]: Will be `data: FormProps`

File: src/Components/payment-form/PaymentForm.tsx [hunk 4]
```diff
@@ -70,7 +69,7 @@ const PaymentForm = () => {
                   />
                   {errors.cardholderName && (
                     <Alert severity="error">
-                      {t(errors.cardholderName.message, {
+                      {t(errors.cardholderName.message as string, {
```
Comment (yeged) [inline src/Components/payment-form/PaymentForm.tsx:72]: You shouldn't need to define as string here.

File: src/Components/payment-form/paymentForm.validation.ts [hunk 5]
```diff
@@ -10,47 +10,47 @@ const schema = yup.object({
     )
```
Comment (yeged) [inline src/Components/payment-form/paymentForm.validation.ts:10]: can you add schema type. Same for Login.

`const schema: Yup.SchemaOf<FormProps> = ...`

Comment (yeged) [reply src/Components/payment-form/paymentForm.validation.ts:10]: if you get error you can use `default('')`

`Yup.string().notRequired().default('')`

File: src/Components/LoginContext.tsx [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+type AppState = typeof initialState
+
+const LoginContext = createContext<{
+  state: AppState,
+  dispatch: React.Dispatch<LoginAction>
+}>({state: initialState, dispatch: () => {}});
+
+type LoginAction = {type: "log in"} | {type: "log out"};
```
Comment (yeged) [inline src/Components/LoginContext.tsx:14]: You can use enum here. If its possible with useReducer
 
```
export enum LoginActionState {
    LOGIN,
    LOGOUT
}

interface LoginAction {
  type: LoginActionState ;
}
```
or another way.

File: src/Components/payment-form/PaymentForm.tsx (hunk 4 above)
Comment (borangoher) [reply src/Components/payment-form/PaymentForm.tsx:72]: Typescript won't compile unless I include ```as string```. If I understand the error message correctly, Typescript expects ```errors.cardholderName.message``` to be a ```string | undefined``` and expects the input to ```t``` to be a ```string | string[]```. So it won't let me pass the argument if I don't narrow the type to just ```string```.

Comment (yeged) [reply src/Components/payment-form/PaymentForm.tsx:72]: Nice research.

I was using `t(errors.amount.message || '')` logical or operator. But we don't want to call t function every time so, we use now `errors.amount ? t(errors.amount.message) : ''` conditional (ternary) operator.

File: src/Components/payment-form/paymentForm.validation.ts (hunk 5 above)
Comment (yeged) [reply src/Components/payment-form/paymentForm.validation.ts:10]: You don't need to add default() if it's required. You wont get error

File: src/api/index.ts [hunk 7]
```diff
@@ -1,4 +1,4 @@
-export const loginUser = async (userData) => {
+export const loginUser = async (userData: object) => {
```
Comment (ozgunbal) [inline src/api/index.ts:1]: don't prefer `object`. this is a generic type. This is actually `FormProps` but you can import as `UserData` in here.
`import { FormProps as UserData } from "../Components/payment-form/paymentForm.constant";`

**Review by ozgunbal — CHANGES_REQUESTED** [review]

**Review by ozgunbal — APPROVED** [review]

**Pull request merged**